src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+                    // will be automatically committed if the user hits the 'invoke completion' 
+                    // without bringing up the completion list.  An item is unique if it was the
+                    // only item to match the text typed so far, and there was at least some text
+                    // typed.  i.e.  if we have "Console.$$" we don't want to commit something
+                    // like "WriteLine" since no filter text has actually been provided.  However,
+                    // if "Console.WriteL$$" is typed, then we do want "WriteLine" to be committed.
+                    for (selectedItemIndex = 0; selectedItemIndex < items.Count; ++selectedItemIndex)
+                    {
+                        if (Equals(items[selectedItemIndex].RoslynCompletionItem, bestItem))
+                            break;
+                    }
+
+                    Debug.Assert(selectedItemIndex < items.Count);
+
+                    bestOrFirstMatchResult = items[selectedItemIndex];
+                    var deduplicatedListCount = matchingItems.Count(r => !r.RoslynCompletionItem.IsPreferredItem());
+                    if (deduplicatedListCount == 1 &&
+                        _filterText.Length > 0)
+                    {
+                        uniqueItem = items[selectedItemIndex].EditorCompletionItem;
+                    }
+                }
+
+                var typedChar = _data.Trigger.Character;
+
+                // Check that it is a filter symbol. We can be called for a non-filter symbol.
+                // If inserting a non-filter character (neither IsPotentialFilterCharacter, nor Helpers.IsFilterCharacter),
+                // we should dismiss completion except cases where this is the first symbol typed for the completion session
+                // (string.IsNullOrEmpty(filterText) or string.Equals(filterText, typeChar.ToString(), StringComparison.OrdinalIgnoreCase)).
+                // In the latter case, we should keep the completion because it was confirmed just before in InitializeCompletion.
+                if (FilterReason == CompletionFilterReason.Insertion &&
+                    !string.IsNullOrEmpty(_filterText) &&
+                    !string.Equals(_filterText, typedChar.ToString(), StringComparison.OrdinalIgnoreCase) &&
+                    !IsPotentialFilterCharacter(typedChar) &&
+                    !Helpers.IsFilterCharacter(bestOrFirstMatchResult.RoslynCompletionItem, typedChar, _filterText))
+                {
+                    return null;
+                }
+
+                var isHardSelection = IsHardSelection(
+                    _filterText, bestOrFirstMatchResult.RoslynCompletionItem, bestOrFirstMatchResult.MatchedFilterText, _hasSuggestedItemOptions);
+
+                var updateSelectionHint = isHardSelection ? UpdateSelectionHint.Selected : UpdateSelectionHint.SoftSelected;
+
+                return new(selectedItemIndex, updateSelectionHint, uniqueItem);
+            }
```

Comment:
note: this method would probably be good split into two.  one that gets the best item, and one that then figures if hte best item works with all this filtering logic.